server/cli/src/graphql/mod.rs

```diff
@@ -0,0 +1,172 @@
+use std::{
+    io::{self, Read},
+    path::PathBuf,
+};
+
+use auth::AUTH_QUERY;
+use reqwest::{multipart, RequestBuilder, Url};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+#[derive(Debug)]
+pub struct Api {
+    url: Url,
+    token: String,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize, Serialize)]
+struct GraphQlResponse {
+    data: Root,
+}
+
+#[allow(dead_code)]
+#[derive(Debug, Deserialize, Serialize)]
+struct Root {
+    root: serde_json::Value,
+}
+
+mod auth;
+pub mod queries_mutations;
+
+#[derive(thiserror::Error, Debug)]
+pub enum ApiError {
+    #[error("Error while sending request to {1}")]
+    SendingRequest(#[source] reqwest::Error, Url),
+    #[error("Error while getting text, status {1:?}")]
+    GettingText(#[source] reqwest::Error, reqwest::StatusCode),
+    #[error("Error parsing gql response: {1}")]
+    ParsingJson(#[source] serde_json::Error, String),
+    #[error("Error validating typename, expected typename {expected_typename}, result: {json}")]
+    ValidatingTypename {
+        expected_typename: String,
+        json: String,
+    },
+    #[error("Error opening file, {1}")]
+    OpeningFile(#[source] io::Error, PathBuf),
+    #[error("Error reading file, {1}")]
+    ReadingFile(#[source] io::Error, PathBuf),
+}
+use service::UploadedFile;
+use ApiError as Error;
+
+impl Api {
+    pub async fn new_with_token(
+        url: Url,
+        username: String,
+        password: String,
+    ) -> Result<Self, Error> {
+        let result = _gql(
+            &url.join("graphql").unwrap(),
+            AUTH_QUERY,
+            serde_json::json! ({
+              "username": username,
+              "password": password,
+            }),
+            None,
+            Some("AuthToken"),
+        )
+        .await?;
+
+        let token = result["token"].as_str().unwrap().to_string();
```

Comment:
extracting and saving token, unwrap should be save as we've checked 'expected_typename' and query is const

Reply:
I agree this is ok